subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junitplatform/JUnitPlatformEnvironmentIntegrationTest.groovy

```diff
@@ -0,0 +1,242 @@
+                public void ok() { }
+            }
+        """
+
+        when:
+        fails('test')
+
+        then:
+        new DefaultTestExecutionResult(testDirectory)
+            .testClassStartsWith('Gradle Test Executor')
+            .assertExecutionFailedWithCause(containsString('consider adding an engine implementation JAR to the classpath'))
+    }
+
+    def "automatically loads junit-platform-launcher from distribution"() {
+        given:
+        buildFile << """
+            testing.suites.test.dependencies {
+                implementation 'org.junit.jupiter:junit-jupiter:${JUNIT_JUPITER_VERSION}'
+            }
+        """
+
+        addClasspathTest("""
+            assertEquals(jars.get(0), "junit-jupiter-params-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(1), "junit-jupiter-engine-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(2), "junit-jupiter-api-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(3), "junit-platform-engine-${JUNIT_PLATFORM_VERSION}.jar");
+            assertEquals(jars.get(4), "junit-platform-commons-${JUNIT_PLATFORM_VERSION}.jar");
+            assertEquals(jars.get(5), "junit-jupiter-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(6), "opentest4j-${OPENTEST4J_VERSION}.jar");
+
+            // The distribtuion-loaded jar can vary in version, since the GE test acceleration
+            // plugins inject their own versions of these libraries during integration tests.
+            // See: https://github.com/gradle/gradle/pull/21494
+            assertTrue(jars.get(7).startsWith("junit-platform-launcher-1."));
+
+            assertEquals(jars.size(), 8);
+        """)
+
+        expect:
+        succeeds "test"
+    }
+
+    def "does not load junit-platform-launcher from distribution when it is on the classpath already"() {
+        given:
+        buildFile << """
+            testing.suites.test.dependencies {
+                implementation 'org.junit.jupiter:junit-jupiter:${JUNIT_JUPITER_VERSION}'
+
+                runtimeOnly 'org.junit.platform:junit-platform-launcher'
+            }
+        """
+
+        addClasspathTest("""
+            assertEquals(jars.get(0), "junit-jupiter-params-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(1), "junit-jupiter-engine-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(2), "junit-jupiter-api-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(3), "junit-platform-launcher-${JUNIT_PLATFORM_VERSION}.jar");
+            assertEquals(jars.get(4), "junit-platform-engine-${JUNIT_PLATFORM_VERSION}.jar");
+            assertEquals(jars.get(5), "junit-platform-commons-${JUNIT_PLATFORM_VERSION}.jar");
+            assertEquals(jars.get(6), "junit-jupiter-${JUNIT_JUPITER_VERSION}.jar");
+            assertEquals(jars.get(7), "opentest4j-${OPENTEST4J_VERSION}.jar");
+            assertEquals(jars.size(), 8);
```

Comment:
💭 This and the previous test might be clearer as just one `contains()` and one `!contains()` assertion - is there a need to test the presence of this exact list of other deps here?

Reply:
Will move to contains-based checking. However, since we don't know the exact version of junit we load from the distribution (due to GE injecting their own version), we can't directly check that our version was loaded. The only way we can be confident about the tests is to verify the entire classpath, including transitive deps, to make sure no extra dependencies were loaded from the distribution which we didnt intend. 

We can get rid of the ordering requirements though